pkg/server/server.go

```diff
@@ -276,6 +307,111 @@ func (h *handler) authenticateEndpoint(w http.ResponseWriter, req *http.Request)
 	})
 }
 
+type NodeNameProvider interface {
+	// Get a node name from instance ID
+	GetNodeName(string) (string, error)
+}
+
+type nodeNameEC2PrivateDNSProvider struct {
+	sess          *session.Session
+	nodeNameCache map[string]string
+	lock          sync.Mutex
+}
+
+func NewNodeNameEC2PrivateDNSProvider(region string, roleARN string) NodeNameProvider {
+	return &nodeNameEC2PrivateDNSProvider{
+		sess:          newSession(region, roleARN),
+		nodeNameCache: make(map[string]string),
+	}
+}
+
+func newSession(region string, roleARN string) *session.Session {
+	// Initial credentials loaded from SDK's default credential chain, such as
+	// the environment, shared credentials (~/.aws/credentials), or EC2 Instance
+	// Role.
+
+	var cfg aws.Config
+	var sess *session.Session
+
+	if region != "" {
+		cfg.Region = aws.String(region)
+	} else {
+		sess = session.Must(session.NewSession())
+		ec2metadata := ec2metadata.New(sess)
+		regionFound, err := ec2metadata.Region()
+		if err != nil {
+			logrus.WithError(err).Fatal("Region not found in config and failed to discover from ec2 instance metadata")
+		}
+		cfg.Region = aws.String(regionFound)
+	}
+
+	sess = session.Must(session.NewSession(&cfg))
+
+	if roleARN != "" {
+		logrus.WithFields(logrus.Fields{
+			"roleARN": roleARN,
+		}).Infof("Using assumed role for EC2 API")
+
+		ap := &stscreds.AssumeRoleProvider{
+			Client:   sts.New(sess),
+			RoleARN:  roleARN,
+			Duration: time.Duration(60) * time.Minute,
+		}
+
+		cfg.Credentials = credentials.NewCredentials(ap)
+		sess = session.Must(session.NewSession(&cfg))
+	}
+	return sess
+}
+
+func (p *nodeNameEC2PrivateDNSProvider) getPrivateDNSName(id string) (string, error) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	name, ok := p.nodeNameCache[id]
+	if ok {
+		return name, nil
+	}
+	return "", errors.New("instance id not found")
+}
+
+func (p *nodeNameEC2PrivateDNSProvider) setPrivateDNSName(id string, privateDNSName string) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	p.nodeNameCache[id] = privateDNSName
+}
+
+// getNodeName looks up the private DNS from the EC2 API
+// and returns a username for the node matching the format that
+// kubelet uses: "system:node:<private-DNS>"
+func (p *nodeNameEC2PrivateDNSProvider) GetNodeName(id string) (string, error) {
+	privateDNSName, err := p.getPrivateDNSName(id)
+	if err == nil {
+		return config.NodeNamePrefix + privateDNSName, nil
+	}
+
+	// Look up instance from EC2 API
+	instanceIds := []*string{&id}
+	ec2Service := ec2.New(p.sess)
+	output, err := ec2Service.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: instanceIds,
+	})
+	if err != nil {
+		return "", err
```

Comment:
nit: add a more explicit message if the api returns an error because instance doesn't exist.

Reply:
That should be caught by the next return, but I made both errors more explicit.